### PR TITLE
bpo-33591: Fix CDLL to accept PathLike objects

### PR DESCRIPTION
**Pull request by mrh1997**
CDLL (and WinDLL, OleDLL, ...) accept now not only str objects but also PathLike objects.

https://bugs.python.org/issue33591

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

Thanks again to your contribution and we look forward to looking at it!

**Comment by mrh1997** [comment]
I signed the CLA (my bugtracker name is "mrh1997").
Please remove the "CLA not signed label"

File: Lib/ctypes/test/test_loading.py [hunk 1]
```diff
@@ -24,10 +24,20 @@ class LoaderTest(unittest.TestCase):
...
-        CDLL(libc_name)
-        CDLL(os.path.basename(libc_name))
+        if libc_name is not None:
+            test_lib = libc_name
+        else:
+            if os.name == "nt":
+                import _ctypes_test
+                test_lib = _ctypes_test.__file__
+            else:
+                self.skipTest('could not find libc')
```
Comment (JelleZijlstra) [inline Lib/ctypes/test/test_loading.py]:
```suggestion
                self.skipTest('could not find library to load')
```
Since we're no longer loading just libc.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
I'm sorry this PR took so long to review. I dug it up from page 57 of the open PR list, but I think it's pretty close to ready.

I do have some comments:
- https://docs.python.org/3.10/library/ctypes.html#ctypes.CDLL should have a "versionchanged" notice saying that starting in version 3.11, PathLike objects are also supported. A similar example note is the one in https://docs.python.org/3.10/library/shutil.html#shutil.move for 3.9. The note should also be added to the other *DLL classes.
- I'd suggest rewording the NEWS entry as:

```
:class:`ctypes.CDLL`, :class:`ctypes.OleDLL`, :class:`ctypes.WinDLL`,
and :class:`ctypes.PyDLL` now accept :term:`path-like objects
<path-like object>` as their ``name`` argument. Patch by Robert Hoelzl.
```

**Review by iritkatriel — DISMISSED** [review]
This has merge conflicts now.

**Comment by arhadthedev** [comment]
@mrh1997 Could you sign the new CLA by clicking `not signed` button in cpython-cla-bot's message, please?

**Comment by arhadthedev** [comment]
For bedevere:

I have made the requested changes; please review again.

**Comment by arhadthedev** [comment]
Non-Linux tests failed with the following error:

```text
======================================================================
ERROR: test_load (test.test_ctypes.test_loading.LoaderTest.test_load)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\Lib\test\test_ctypes\test_loading.py", line 44, in test_load
    CDLL(CTypesTestPathLikeCls())
  File "D:\a\cpython\cpython\Lib\ctypes\__init__.py", line 367, in __init__
    if '/' in name or '\\' in name:
       ^^^^^^^^^^^
TypeError: argument of type 'CTypesTestPathLikeCls' is not iterable
```

I'm investigating it.

**Comment by arhadthedev** [comment]
Fixed; there were two missed lines that continued to use `name` function argument directly instead of stringified `self._name`.

File: Lib/ctypes/__init__.py [hunk 2]
```diff
@@ -356,15 +356,15 @@ def __init__(self, name, mode=DEFAULT_MODE, handle=None,
                archive(member) syntax for dlopen(), and the mode is adjusted.
                Otherwise, name is presented to dlopen() as a file argument.
             """
-            if name and name.endswith(")") and ".a(" in name:
```
Comment (kumaraditya303) [inline Lib/ctypes/__init__.py]: It would be cleaner to just reassign `name to self._name` and not change these.

Comment (arhadthedev) [reply Lib/ctypes/__init__.py]: Done.

File: Lib/ctypes/__init__.py [hunk 3]
```diff
@@ -344,7 +344,8 @@ def __init__(self, name, mode=DEFAULT_MODE, handle=None,
                  use_errno=False,
                  use_last_error=False,
                  winmode=None):
-        self._name = name
+        if name:
+            name = _os.fspath(name)
```
Comment (kumaraditya303) [inline Lib/ctypes/__init__.py:348]: Why not just assign name here itself to self? You don't need to change it everywhere.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Lib/ctypes/__init__.py (hunk 3 above)
Comment (arhadthedev) [reply Lib/ctypes/__init__.py:348]: Thank you, modified. Initially I wanted to deduplicate `name` and `self._name` but I agree that it makes the diff unnecessarily complicated.

**Comment by arhadthedev** [comment]
I have made the requested changes; please review again.

**Review by kumaraditya303 — APPROVED** [review]
